File: apps/bridge/bridge/embed/cohere_rerank.py
from __future__ import annotations

import httpx

from bridge.embed.protocols import RerankResult


class CohereStyleReranker:
    """
    Rerank provider for the Cohere-style /rerank endpoint.

    Cohere defined this wire format; Jina, SiliconFlow, and Voyage all use it:

        POST {base_url}/rerank
        {"model": "...", "query": "...", "documents": [...], "top_n": N}

        Response:
        {"results": [{"index": N, "relevance_score": 0.9, "document": {...}}]}
    """

    def __init__(
        self,
        model: str,
        api_key: str,
        base_url: str,
        timeout: float = 30.0,
    ) -> None:
        self._model = model
        # Ensure /rerank path is absolute — strip trailing slash from base_url.
        self._rerank_url = base_url.rstrip("/") + "/rerank"
        self._headers = {
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        }
        self._timeout = timeout

    async def rerank(
        self,
        query: str,
        docs: list[str],
        top_k: int,
    ) -> list[RerankResult]:
        if not query or not docs:
            raise ValueError("query and docs must be non-empty")

        async with httpx.AsyncClient(timeout=self._timeout) as client:
            resp = await client.post(
                self._rerank_url,
                headers=self._headers,
                json={
                    "model": self._model,
                    "query": query,
                    "documents": docs,
                    "top_n": top_k,
                },
            )
            resp.raise_for_status()
            data = resp.json()

        results = [
            RerankResult(
                index=r["index"],
                score=float(r.get("relevance_score", r.get("score", 0.0))),
            )
            for r in data.get("results", [])
        ]
        results.sort(key=lambda r: r.score, reverse=True)
        return results[:top_k]

    async def health(self) -> bool:
        try:
            await self.rerank("test", ["test document"], top_k=1)
            return True
        except Exception:
            return False

    def __repr__(self) -> str:
        return f"CohereStyleReranker(model={self._model!r}, url={self._rerank_url!r})"

File: packages/ebd-client/src/adapters/cohere-rerank.ts
import type { RerankProviderConfig, RerankResponse } from '../types.js';
import type { RerankAdapter } from './base.js';

/**
 * Cohere-style rerank adapter.
 * Compatible with: Cohere, SiliconFlow, Jina, Voyage.
 *
 * Wire format:
 *   POST {baseUrl}/rerank
 *   { model, query, documents: string[], top_n }
 *   → { results: [{ index, relevance_score }] }
 */
function normalizeBaseUrl(baseUrl: string | undefined): string {
  const clean = baseUrl?.trim().replace(/\/+$/, '');
  if (!clean) {
    throw new Error('cohere-rerank: baseUrl is required for the Cohere rerank protocol');
  }
  return clean;
}

export class CohereRerankAdapter implements RerankAdapter {
  private readonly config: RerankProviderConfig;

  constructor(config: RerankProviderConfig) {
    this.config = config;
  }

  async rerank(
    query: string,
    documents: string[],
    topK: number,
    model: string,
  ): Promise<RerankResponse> {
    const baseUrl = normalizeBaseUrl(this.config.baseUrl);
    const res = await fetch(`${baseUrl}/rerank`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.config.apiKey}`,
      },
      body: JSON.stringify({ model, query, documents, top_n: topK }),
    });

    if (!res.ok) {
      const text = await res.text().catch(() => '');
      throw new Error(`cohere-rerank HTTP ${res.status}: ${text}`);
    }

    const data = await res.json() as {
      results: { index: number; relevance_score?: number; score?: number }[];
    };

    const results = data.results
      .map(r => ({
        index: r.index,
        score: r.relevance_score ?? r.score ?? 0,
      }))
      .sort((a, b) => b.score - a.score)
      .slice(0, topK);

    return { results };
  }
}
